### PR TITLE
Splits sites object in the app state

### LINKED ISSUE #1646: Wrong favicon is shown for bookmark
captura de ecran din 2016 05 09 la 18 13 07

Sometimes favicons get mixed up. In the above screenshot, the favicon for feedly was green while I was reading the RSS feeds. Then I clicked the twitter link and the feedly favicon changed to that of twitter.

### LINKED ISSUE #1856: Removing sites from history - `siteUtil` functionality
Coming to implement #1458 , I needed a way to remove a site from history.
Currently, on the `siteUtil` state, there is a function called `removeSite`, for convenience, here is it:

``` javascript
module.exports.removeSite = function (sites, siteDetail, tag) {
  const index = module.exports.getSiteIndex(sites, siteDetail, tag)
  if (index === -1) {
    return sites
  }
  const tags = sites.getIn([index, 'tags'])
  return sites.setIn([index, 'tags'], tags.toSet().remove(tag).toList())
}
```

It currently only removes tags from sites in the list (such as `bookmark` tag, which is the main thing this function is used for now, removing bookmarks).

sites without tags are sites saved in history, so there are these possibilities I had in mind, but I wanted someone who has a better grasp of the code to tell me which seems to be the best:
- Add the functionality of removing a site from history to this function, when there is not `tag` specified. This effectively makes the function have two responsibilities.
- Add a new function `removeHistorySite` that removes sites without tags from the map, but have the chance of being confusing in the future (naming)
- Refactor the current `removeSite` name to be `removeTag`, and make the `removeSite` function removes sites without tags from the map.

Sorry for being a hard to deal with contributor (:

### LINKED ISSUE #2655: Bookmark name of a file does not appear in the autocomplete list
**Describe the issue you encountered:** If you bookmark a URL to a file and set a title to the bookmark, only the URL is displayed in the autocomplete list.

**Expected behavior:** The title also should be displayed.
- Platform (Win7, 8, 10? macOS? Linux distro?): OS X
- Brave Version: 0.11.1
- Steps to reproduce:
  1. Bookmark http://speedtest.ftp.otenet.gr/files/test100k.db with the title `test`
  2. Enter `test`
- Screenshot if needed:
  screenshot 2016-07-22 10 43 13
- Any related issues: #2271

### LINKED ISSUE #2771: Bookmark folder is created left of bookmark item, not right
**Describe the issue you encountered:** Bookmark folder is created left of a bookmarked item. It's the same after moving `brave` folder in `AppData\Roaming`.

**Expected behavior:** It should be created right of the item
- Platform (Win7, 8, 10? macOS? Linux distro?): Windows 7 64 bit
- Brave Version: 0.11.2 Beta1
- Steps to reproduce:
  1. Bookmark https://github.com from the bookmark toolbar
  2. Create a bookmark folder
- Screenshot if needed:
- Any related issues: #2768

### LINKED ISSUE #3646: Bookmarks title changes when navigating around the page
**Did you search for similar issues before submitting this one?**
Yes.

**Describe the issue you encountered:**
It seems like the title in the bookmarks bar changes as you navigate around the site. For example, I bookmarked twitter.com so the title says Twitter. When I navigate to twitter's notifications tab the title changes in my bookmarks bar to Twitter / Notifications.

**Expected behavior:**

It should stay the same.
- Platform (Win7, 8, 10? macOS? Linux distro?):

macOS
- Brave Version:

0.11.6
- Steps to reproduce:
  1. Bookmark twitter.com
  2. Make sure bookmarks bar is showing
  3. Visit twitter.com and then go to notifications
  4. Notice the title changes, causing the bookmarks to visibly move around
- Screenshot if needed:

bug
- Any related issues:

### LINKED ISSUE #3694: Add automated test for making sure bookmark custom title appears on bookmark toolbar
This test should add a bookmark with a custom title and make sure it shows that custom title.

Code to modify for the test is here:
`test/components/bookmarksTest.js`

Tests for each React component (UI element) is in `test/components` and it maps to `js/components` code.

You'll notice this test is currently split into 2 parts, bookmarking titles of pages with titles, and sites without titles. So you'll add 2 new tests in this issue, one for each.

Most of how tests work is using the DOM inspector with Shift+F8 dev tools to find what the css selector is for a certain DOM element.  You can test your CSS selector to make sure it selects the correct element just by opening the browser manually, opening browser dev tools and in the console typing document.querySelector('your-selector-here').

Always try to make sure individual `it` blocks in the test do not depend on each other.  The `describe`'s before block usually sets up the test, and then each `it` tests something about it.

The best way to figure out details of the test is just to read other tests but we have some good documentation including debugging information here: https://github.com/brave/browser-laptop/blob/master/docs/tests.md

### LINKED ISSUE #4224: Hidden history entries appear by deleting the latest entry
**Describe the issue you encountered:** Hidden history entries appear by deleting the latest entry.

history

5:50:14 PM -> 5:50:22 PM -> 5:51:04 PM -> 5:51:06 PM

I am not sure of steps to reproduce the issue.

**Expected behavior:** No history entries should be hidden.
- Platform (Win7, 8, 10? macOS? Linux distro?): Windows 10
- Brave Version: 0.12.1
- Any related issues:

### LINKED ISSUE #4260: Edited information of bookmark does not appear on bookmark dialog via star icon
**Describe the issue you encountered:** The edited information (title and URL) of bookmark does not appear on bookmark dialog opened via star icon. This issue does not happen if you click the icon on the toolbar to select Edit Bookmark.

**Expected behavior:** The edited information (title and url) of bookmark should be filled in these input fields.
- Platform (Win7, 8, 10? macOS? Linux distro?): Windows 10
- Brave Version: 0.12.2 RC3
- Steps to reproduce:
  1. Open #4255
  2. Click the star icon next to the url bar
  3. Edit Title and click Save
  4. Click the star icon again
  5. The initial title appears instead of the edited one
- Any related issues:

Please set a milestone if necessary.

### LINKED ISSUE #4833: Bookmark title dynamically changed
**Did you search for similar issues before submitting this one?**
No
**Describe the issue you encountered:**
Bookmark title dynamically changed when its open a new link with different title. For example in my case is twitch. I bookmark https://www.twitch.tv/directory/following its title Twitch - following and when I opened someone's stream page it also changed my bookmark title for twith. I dont know if this is actually should working like that, but if yes I would like to have some options like "Allow/disallow dynamically change on bookmark bar".

Sorry for bad grammar

**Expected behavior:**
It should be not change the title whenever we opened another link.
- Platform (Win7, 8, 10? macOS? Linux distro?):
  Win 7
- Brave Version:
  Latest version
- Steps to reproduce:
  1.
  2.
  3.
- Screenshot if needed:
- Any related issues:

### LINKED ISSUE #4868: Rebookmarking from context menu does not reset bookmark title
**Describe the issue you encountered:** Rebookmarking does not reset the modified bookmark title to the default one.

**Expected behavior:** It should reset the bookmark title.

- Platform (Win7, 8, 10? macOS? Linux distro?): Windows 10

- Brave Version: 0.12.5 RC3

- Steps to reproduce:
  1. Clear brave folder
  2. Open http://slashdot.org in a new tab
  3. "Add Bookmark" from context menu on Bookmark Toolbar with a modified title, like `test`
  4. "Add Bookmark" from context menu on Bookmark Toolbar again with the default title, which is `Slashdot: ...`
  5. The title remains `test`

- Any related issues:

### LINKED ISSUE #4929: Bookmark title automatically changes on the bookmark form
**Describe the issue you encountered:** Bookmark title automatically changes on the bookmark form, if you edit the bookmarked URL, open it, and edit the bookmark again.

**Expected behavior:** It should either preserve the original title or renew it based on the title of the page of the edited URL.
- Platform (Win7, 8, 10? macOS? Linux distro?): OS X
- Brave Version: https://github.com/brave/browser-laptop/tree/fc14b2ed23e23dd90d518766321324a1e79a2e90
- Steps to reproduce:
  1. Bookmark `https://github.com/notifications`
  2. Change Location via "Edit Bookmark" on the bookmark toolbar to `https://github.com/notificationsa`, which should lead to 404 error
  3. Click the bookmark
  4. "Edit Bookmark"
  5. The title was changed to `Page not found · GitHub`
- Any related issues: #4897

### LINKED ISSUE #5072: about:history omits recent history
about:history only displays browsing history more than ~22h57m old. History data seems correct as missing browsing history does appear once sufficient time has passed but the display should include everything upto the most recent browsing history.

Ubuntu 16.10
Brave: 0.12.5
Electron: 1.4.14
libchromiumcontent: 53.0.2785.143
V8: 5.3.332.47
Node.js: 6.5.0
Update Channel: dev

### LINKED ISSUE #5382: clicking entries on about:history is laggy
**Describe the issue you encountered:** clicking entries on about:history is laggy. You have to wait ~0.5s until the entry you click is actually selected.

**Expected behavior:** it should be selected instantly

- Platform (Win7, 8, 10? macOS? Linux distro?): macOS

- Brave Version: master branch

- Steps to reproduce:
    1. Open about:history
    2. Click an entry
    3. Wait ~0.5s until the entry is selected

- Any related issues:

### LINKED ISSUE #5699: Last Visited on Bookmark Manager is not recorded for newly created bookmarks
## Test plan

https://github.com/brave/browser-laptop/pull/10136#issue-245629798

----

**Describe the issue you encountered:** Last Visited on Bookmark Manager is not recorded for newly created bookmarks

**Expected behavior:** It should be recorded

- Platform (Win7, 8, 10? macOS? Linux distro?): macOS

- Brave Version: https://github.com/brave/browser-laptop/tree/e1b985d240ff72945bfc76e5084da0cf373ae092

- Steps to reproduce:
    1. Open about:bookmarks
    2. Create a bookmark 1 (title: null, location: `brave.com`)
    2. Create a bookmark 2 (title: null, location: `https://brave.com`)
    3. Click both
    4. Reopen the bookmark manager

- Any related issues:

### LINKED ISSUE #6104: Address bar should autocomplete from custom bookmark titles
**Did you search for similar issues before submitting this one?**
Yes

**Describe the issue you encountered:**
I sometimes bookmark URLs I'll know I'll need again. When I start typing their names (not just addresses) Brave should match it and suggest a completion.

**Expected behavior:**
Brave doesn't match anything.

- Platform (Win7, 8, 10? macOS? Linux distro?):
Mac, Windows

- Brave Version:
Brave	0.12.13
Muon	1.4.30
libchromiumcontent	53.0.2785.143
V8	5.3.332.47
Node.js	6.5.0
Update Channel	dev
os.platform	darwin
os.release	16.3.0
os.arch	x64

- Steps to reproduce:
    1. Bookmark an URL with a distinctive title, e.g. "testmark"
    2. Type "testmark" into the URL bar

### LINKED ISSUE #6108: Bookmark title on bookmark toolbar is renamed
**Describe the issue you encountered:**
Bookmark title on the bookmark toolbar is renamed after clicking the link to the notification page on GitHub.

**Expected behavior:**
The bookmark title should not be changed.

- Platform (Win7, 8, 10? macOS? Linux distro?): macOS

- Brave Version: master branch

- Steps to reproduce:
    1. Bookmark the notification page on GitHub
    2. Click the notification link -> the bookmark title is changed

- Screenshot if needed:
bookmark

- Any related issues:

### LINKED ISSUE #6585: Remove customTitle from siteDetail object
Discussion starts here https://github.com/brave/browser-laptop/issues/6108#issuecomment-265934841

> **To keep things simple, I'd like to propose that we eliminate custom title**.
> - When a bookmark is added for a page, it will default to the title of the page
> - Once set, the title for that bookmark will never be updated again

Removing this should resolve the following issues:
https://github.com/brave/browser-laptop/issues/6104
https://github.com/brave/browser-laptop/issues/6108
https://github.com/brave/browser-laptop/issues/3694

There are likely more which have this as the root cause (we can search for issues where titles change)

### LINKED ISSUE #8022: Bookmark moved to a folder on the toolbar cannot be removed via the star button
**Describe the issue you encountered:** Bookmark in a folder on the toolbar cannot be removed via the star button, unless you select the bookmark folder into which you moved the bookmark. By default it is set to "bookmark toolbar" and not overwritten.

I am not sure if this is a regression.

- Platform (Win7, 8, 10? macOS? Linux distro?): Windows 10

- Brave Version (revision SHA): 0.14.1 RC1

- Steps to reproduce:
    1. Open a page
    2. Bookmark it 
    3. Add a bookmark folder
    4. Move the bookmark into the folder
    5. Click the star button on the URL bar
    6. Click "Remove" -> bookmark is not removed
    7. Click the star
    8. **Choose the folder** and click "Remove" -> bookmark is removed
    
- Actual result: the bookmark folder is not overwritten after moving the bookmark to another folder from "bookmark toolbar"

- Expected result: it should be overwritten

- Extra QA steps:
    1.
    2.
    3.

- Screenshot if needed:

- Any related issues:

### LINKED ISSUE #9301: Changing bookmark location causes duplicate bookmarks
- Did you search for similar issues before submitting this one?
Yes, didnt find any issues with bookmark tag

- Describe the issue you encountered:
Changing bookmark location causes duplicate bookmarks

- Platform (Win7, 8, 10? macOS? Linux distro?):
All

- Brave Version (revision SHA):
Brave	0.16.3
rev	9f7c3b7
Muon	3.2.102

- Steps to reproduce:
    1. Create a bookmark folder 
    2. Add a bookmark and add under the created folder
    3. Visit the site by clicking on the bookmark
    4. Edit the bookmark in the URL and change the location to `Bookmark Toolbar` 
    5. Bookmark is duplicated under the folder and the toolbar
    
- Actual result:
Changing bookmark location causes duplicate bookmarks

- Expected result:
Should only have one bookmark and not duplicate. Behaviour is different from Chrome

- Will the steps above reproduce in a fresh profile? If not what other info can be added?
Yes

- Is this an issue in the currently released version?
Yes, 
Edit: Have checked it from 0.14.1 and its the same issue

- Can this issue be consistently reproduced?
Yes

- Extra QA steps:
    1.
    2.
    3.

- Screenshot if needed:
duplicatebookmarks

- Any related issues:

### LINKED ISSUE #9326: Custom bookmark title is not being displayed in suggestions
- Did you search for similar issues before submitting this one?
Yes

- Describe the issue you encountered:
While verifying another issue, it was discovered that custom bookmark titles aren't being displayed in the urlbar suggestions menu

- Platform (Win7, 8, 10? macOS? Linux distro?):
Discovered on OS X

- Brave Version (revision SHA):
0.16.4

- Steps to reproduce:
    0. Make sure 'Show bookmark matches' is enabled on about:preferences#search
    1. Create a bookmark with a custom title
    2. Open a new tab and type in the custom title.
    3. Check the suggestion content
    
- Actual result:
The suggestions do not display the custom title

- Expected result:
The suggestions should display custom titles if they exist

- Will the steps above reproduce in a fresh profile? If not what other info can be added?
Yes

- Is this an issue in the currently released version?
Yes

- Can this issue be consistently reproduced?
Yes

- Extra QA steps:
    1.
    2.
    3.

- Screenshot if needed:

- Any related issues:

### LINKED ISSUE #9978: Split sites into separate entities.
**Describe the issue you encountered:**
Split `sites` entity in `appState` into:

- bookmarks
- bookmarkFolders
- visitedSites
- pinnedSites

### LINKED ISSUE #10026: Bookmarks toolbar : imported sorted bookmarks are represented in wacky order
- Did you search for similar issues before submitting this one?
Yes

- Describe the issue you encountered:
Importing a small set of sorted bookmarks ( < 50 ) the bookmarks menu representation is in order but the bookmarks toolbar menu is not.

- Platform (Win7, 8, 10? macOS? Linux distro?):
MacOS

- Brave Version (revision SHA):
7b31653eb555ba8f00395cd593f9b948ada296fe

- Steps to reproduce:
    1. Import 
[brave_sorted_simple.html.zip](https://github.com/brave/browser-laptop/files/1153889/brave_sorted_simple.html.zip)
    2. Check representations in UI

    
- Actual result:
Bookmarks toolbar menu isn't sorted

- Expected result:
Bookmarks toolbar menu should be sorted

- Will the steps above reproduce in a fresh profile? If not what other info can be added?
Yes

- Is this an issue in the currently released version?
Likely

- Can this issue be consistently reproduced?
Yes

- Extra QA steps:
    1.
    2.
    3.

- Screenshot if needed:
screen shot 2017-07-17 at 3 21 28 pm
vs
screen shot 2017-07-17 at 3 21 39 pm

- Any related issues:

### PR DESCRIPTION
**Pull request by NejcZdovc**
Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/browser-laptop/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request [as needed](https://github.com/brave/browser-laptop/wiki/Pull-request-process).

Resolves #1646 (Wrong favicon is shown for bookmark)
Resolves #1856 (Removing sites from history - `siteUtil` functionality)
Resolves #2655 (Bookmark name of a file does not appear in the autocomplete list)
Resolves #2771 (Bookmark folder is created left of bookmark item, not right)
Resolves #3646 (Bookmarks title changes when navigating around the page)
Resolves #3694 (Add automated test for making sure bookmark custom title appears on bookmark toolbar)
Resolves #4224 (Hidden history entries appear by deleting the latest entry)
Resolves #4260 (Edited information of bookmark does not appear on bookmark dialog via star icon)
Resolves #4833 (Bookmark title dynamically changed)
Resolves #4868 (Rebookmarking from context menu does not reset bookmark title)
Resolves #4929 (Bookmark title automatically changes on the bookmark form)
Resolves #5072 (about:history omits recent history)
Resolves #5382 (clicking entries on about:history is laggy)
Resolves #5699 (Last Visited on Bookmark Manager is not recorded for newly created bookmarks)
Resolves #6104 (Address bar should autocomplete from custom bookmark titles)
Resolves #6108 (Bookmark title on bookmark toolbar is renamed)
Resolves #6585 (Remove customTitle from siteDetail object)
Resolves #8022 (Bookmark moved to a folder on the toolbar cannot be removed via the star button)
Resolves #9301 (Changing bookmark location causes duplicate bookmarks)
Resolves #9326 (Custom bookmark title is not being displayed in suggestions) 
Resolves #9978 (Split sites into separate entities)
Resolves #10026 (Bookmarks toolbar : imported sorted bookmarks are represented in wacky order)

Auditors: @bbondy @ayumi @diracdeltas @bsclifton @darkdh 

Test Plan:
Run general regression tests first: https://github.com/brave/browser-laptop/wiki/Manual-Tests

1) Bookmark
- try to add a bookmark
2) Bookmark
- try to add a bookmark, but into the second level folder
3) Bookmark folder
- try to add a bookmark folder
4) Bookmark
- try to edit a bookmark
5) Bookmark folder
- try to add a bookmark folder
6) Bookmark
- try to remove a bookmark
7) Bookmark folder
- try to remove a bookmark folder
8) Bookmark
- try to add a bookmark into a folder (with a right click on a folder)
- folder should be preselected
9) Bookmark folder
- try to add a bookmark folder into a folder (with a right click on a folder)
- folder should be preselected
10) Navigation bar
- try to bookmark a page via navigation bar star
11) Navigation bar
- try to edit a bookmark via navigation bar star
12) Bookmark manager
- try to add a bookmark folder from a bookmark manager  (list, folder tree and button)
13) Bookmark manager
- try to add a bookmark from a bookmark manager  (list, folder tree and button)
14) Bookmark manager
- try to edit a bookmark from a bookmark manager (list and folder tree)
15) Bookmark manager
- try to edit a bookmark folder from a bookmark manager (list and folder tree)
16) Bookmark link
- try to bookmark a link (right click on a link in the page)
17) Bookmark manager
- try to dnd bookmark into another folder
18) Bookmark manager
- try to dnd multiple bookmarks into another folder
19) Bookmark manager
- try to dnd folder into another folder
20) Bookmark manager
- try to reorder bookmarks in bookmark manager
21) Bookmark toolbar
- try to dnd bookmark into the folder
22) Bookmark toolbar
- try to dnd folder into another folder
23) Bookmark toolbar
- try to to reorder bookmarks with dnd (try it with folders as well)
24) History manager
- check if history is successfully added
25) History manager
- right click on a history item
- try to delete it
26) Pinned tabs
- try to pin a site
- restart the browser
- check if pinned tab is still there
27) Importer
- try to import HTML file
27) Importer
- try to import from chrome
27) Importer
- try to import from FF
28) Exporter
- check if export is working
29) Url bar suggestions
- check if bookmark is shown in suggestion list
30) Url bar suggestions
- check if history site is shown in suggestion list
31) Ledger
- enabled ledger
- visit some site
- check if site is added to the ledger
32) New tab
- check if top sites are displayed correctly 
33) New tab
- check if you can remove top site
34) New tab
- check if you can pin top site
35) New tab
- check if you can bookmark and edit bookmark already bookmarked top site
36) New tab
- check if you can reorder top sites

Reviewer Checklist:

Tests

- [ ] Adequate test coverage exists to prevent regressions
- [ ] Tests should be independent and work correctly when run individually or as a suite [ref](https://github.com/brave/browser-laptop/wiki/Code-Guidelines#test-dependencies)
- [ ] New files have MPL2 license header

### CODE REVIEW

**Comment by NejcZdovc** [comment]
### Pinned sites
- [x] create utils
- [ ] create tests for util file
- [x] create upgrade script

### Default sites
- [x] remove from state
- [x] refactor newTab to use topSites directly

### History sites (current sites object)
- [x] create utils
- [ ] create index-order table do we need?
- [ ] create tests for util file
- [x] create upgrade script
- [x] fix about:history

### Bookmarks sites
- [x] create utils
- [x] remove customTitle
- [x] create index-order table
- [ ] create tests for util file
- [x] create upgrade script
- [x] fix about:bookmarks
- [x] try to add a bookmark (star icon) while page is still loading
- [x] new tab bookmark is not working
- [x] changing bookmark folder location is not working for already bookmarked item
- [x] right click on a bookmark and clicking add site deletes all bookmarks

### Bookmark folder sites
- [x] create utils
- [x] remove customTitle
- [x] create index-order table
- [ ] create tests for util file
- [x] create upgrade script

### General
- [x] check how sites (no tags) array is working in other places
- [ ] update sync related code, ask for help from @ayumi @diracdeltas 
- [x] update import related code, ask for help from  @darkdh 
- [x] update suggestions related code, ask for help from  @bbondy  
- [x] alphabetize state docs
- [ ] keep only general functions in `js/state/siteUtil.js`
- [x] remove `app/common/state/siteState.js`
- [x] remove `app/browser/reducers/sitesReducer.js`
- [x] make sure that import order is preserved
- [x] go through all `getSiteKey` calls and see where we can remove them, because key is now inside bookmarks object

## Questions
- [x] can we remove site from `sites` when we pin it?
- [x] should we be saving keys in string format? Otherwise every time you want to use key in `.get()`, you need to add `toString()`
- [ ] we have last visited time in about:bookmarks. Is this needed? @bradleyrichter if so we need to to update bookmark entry when we add history entry.

## Known issues
- [x] improve import, because now it's taking too long to finish (maybe the problem is that we are calling action inside action)
- [x] if you add a site via dialog, we need to add favicon somehow if it's not in the history
- [x] import from chrome is not working
- [ ] sync is not working
- [x] site cache is broken
- [x] if you delete a history item and then bookmark active frame you will lose favicon and theme color
- [x] when importing from chrome last accessed time is not imported correctly
- [x] add sort values for upgrade process
- [x] bookmarks don't get favicon if you bookmark top sites on a fresh profile
- [x] bookmark star status is not displayed correctly on a new tab
- [x] fix dnd for top sites in new tab
- [x] more bookmarks doesn't work on bookmark toolbar

## Tests
- [ ] app/browser/reducers/bookmarkFoldersReducer.js
- [ ] app/browser/reducers/bookmarksReducer.js
- [ ] app/browser/reducers/historyReducer.js
- [ ] app/browser/reducers/pinnedSitesReducer.js
- [ ] app/browser/reducers/sitesReducer.js (this file was removed, move things around)
- [ ] app/common/cache/bookmarkLocationCache.js (before it was siteCache.js)
- [ ] app/common/cache/bookmarkOrderCache.js
- [ ] app/common/lib/bookmarkFoldersUtil.js
- [ ] app/common/lib/bookmarkUtil.js
- [ ] app/common/lib/historyUtil.js
- [ ] app/common/lib/pinnedSitesUtil.js
- [ ] app/common/state/aboutHistoryState.js
- [ ] app/common/state/aboutNewTabState.js
- [ ] app/common/state/bookmarkFoldersState.js
- [ ] app/common/state/bookmarksState.js
- [ ] app/common/state/historyState.js
- [ ] app/common/state/pinnedSitesState.js
- [ ] app/common/state/tabState.js
- [ ] js/state/siteUtil.js (test file needs to be updated)
- [ ] js/stores/windowStore.js

## Possible fixes (check if this issues are fixed after done)
- #5382
- #5072 
- #2655

**Comment by bsclifton** [comment]
Updated parent issue so that the titles for the issues are shown (instead of just the issue numbers)
 😄

File: js/state/syncUtil.js [hunk 1]
```diff
@@ -485,20 +571,22 @@ module.exports.createSiteData = (site, appState) => {
...
       folderToObjectMap[site.parentFolderId] ||
       findOrCreateFolderObjectId(site.parentFolderId, appState)
     value = {
       site: siteData,
-      isFolder: siteUtil.isFolder(immutableSite),
+      isFolder,
       hideInToolbar: site.parentFolderId === -1,
       parentFolderObjectId
     }
   } else if (siteUtil.isHistoryEntry(immutableSite)) {
```
Comment (NejcZdovc) [inline js/state/syncUtil.js]: @ayumi can we remove this function? This function is not relevant anymore. So if we needed it we need to write a new one in historyUtil file.

File: js/state/syncUtil.js [hunk 2]
```diff
@@ -219,27 +233,95 @@ const applySyncRecord = (record) => {
...
+      const siteData = module.exports.getSiteDataFromRecord(record, appState, records)
+      if (shouldAddRecord(record)) {
+        if (record.bookmark.isFolder) {
+          bufferAppAction(appActions.addBookmarkFolder, siteData)
+        } else {
+          bufferAppAction(appActions.addBookmark, siteData)
+        }
+      } else if (shouldRemoveRecord(record)) {
+        const folderKey = bookmarkFoldersUtil.getKey(siteData)
+        // TODO: removeBookmarkFolder doesn't support List removal
```
Comment (NejcZdovc) [inline js/state/syncUtil.js]: @ayumi now it does :smiley:

**Comment by NejcZdovc** [comment]
@bbondy @bsclifton even though that not all tests are done, it would be good if you can check out this PR and give some initial feedbacks. This PR is done, just missing some tests

**Comment by bsclifton** [comment]
Wow- performance is GREAT with this branch 😻

File: app/browser/tabs.js [hunk 3]
```diff
@@ -134,21 +137,12 @@ ipcMain.on(messages.ABOUT_COMPONENT_INITIALIZED, (e) => {
...
-    }
-  })
-  const bookmarks = bookmarkSites.toList().toJS()
-  const bookmarkFolders = bookmarkFolderSites.toList().toJS()
-  return {bookmarks, bookmarkFolders}
+const getBookmarksData = (state) => {
+  return {
+    bookmarks: bookmarksState.getBookmarks(state).toJS(),
+    bookmarkFolders: bookmarkFoldersState.getFolders(state).toJS(),
+    bookmarkOrder: bookmarkOrderCache.getOrderCache(state).toJS()
```
Comment (bsclifton) [inline app/browser/tabs.js]: ex: maybe we can do toJS() once

**Comment by bsclifton** [comment]
Testing and I noticed this bug:
Dragging a bookmark onto the bookmarks toolbar didn't trigger a re-render. If I open a new window, that window has bookmark in the place I dragged to.

When bookmarking, I'm dragging the lock icon in the URL bar onto the bookmarks toolbar and putting it in the first available position

**Comment by NejcZdovc** [comment]
Note for me: @bsclifton bug is not working because `bookmarkUtil.js` line 78 is not having order anymore and when we change order this if will not be failing.

File: app/browser/tabs.js (hunk 3 above)
Comment (NejcZdovc) [inline app/browser/tabs.js]: done

File: app/common/cache/bookmarkLocationCache.js [hunk 4]
```diff
@@ -27,12 +14,40 @@ const normalizeLocation = (location) => {
...
+
+  const splitKey = siteKey.split('|', 2)
+  if (typeof splitKey[0] === 'string' && typeof splitKey[1] === 'string') {
+    return splitKey[0]
+  }
+  return null
+}
+
+const generateCache = (state) => {
+  const cache = state.getIn(['cache', 'bookmarkLocation']) || Immutable.Map()
```
Comment (NejcZdovc) [inline app/common/cache/bookmarkLocationCache.js]: we need it here

**Comment by ayumi** [comment]
:+1:

**Pull request merged**